src/arch/riscv/kernel/boot.c

```diff
@@ -159,6 +162,11 @@ BOOT_CODE static bool_t try_init_kernel_secondary_core(word_t hart_id, word_t co
     init_cpu();
     NODE_LOCK_SYS;
 
+#if defined(CONFIG_KERNEL_MCS) && defined(CONFIG_DEBUG_BUILD)
+    clock_sync_test[getCurrentCPUIndex()] = getCurrentTime();
+    /* Make sure clock_sync_test is set before ksNumCPUs */
+    __atomic_signal_fence(__ATOMIC_ACQ_REL);
```

Comment:
If I get the documentation right,  `__atomic_signal_fence()`does not guarantee a pipeline barrier instruction is created, it's just about the compiler's instruction generation order. So this must be `__atomic_thread_fence(__ATOMIC_ACQ_REL)` then, as we have to ensure the new value in `clock_sync_test` is visible before (or at the same time) of the change in `ksNumCPUs`.

Another observation is, that currently the check runs once we have obtained the BKL (from `NODE_LOCK_SYS` above) I think we could run the clock sync test even before trying to get the BKL, because it's completely unrelated to it and intended to run in parallel on all cores. Having the BKL It does not harm, just adds some more delay before the core's clock value is available.

Reply:
I think you are right about that `__atomic_signal_fence` should be replaced with `__atomic_thread_fence`. However, this should probably also happen for the existing call. I'll replace all calls to be on the safe side.

Currently the check only runs on core 0 without the BKL held, the other cores only store their timestamp (with the BLK held, but that does not matter). The test is done while core 0 is busy looping and waiting for the other cores to finish startup. The time between changing `ksNumCPUs` and for that change to propagate across nodes is the only measuring overhead.

I tried doing the test the other way round, with core 0 making a timestamp and then doing the compare in the launched cores, but that adds too much cpu init overhead for correct time measurements.